queue/example_connection_pool_test.go

```diff
@@ -0,0 +1,148 @@
+	if role != connection_pool.MasterRole {
+		return nil
+	}
+
+	// We need to register a shared session only once.
+	h.registerMutex.Lock()
+
+	if h.err != nil {
+		h.registerMutex.Unlock()
+		return h.err
+	}
+
+	q := queue.New(conn, h.name)
+	if !h.registered {
+		opts := queue.CfgOpts{InReplicaset: true, Ttr: 60 * time.Second}
+		if h.err = q.Cfg(opts); h.err == nil {
+			if h.uuid, h.err = q.Identify(nil); h.err == nil {
+				h.err = q.Create(h.cfg)
+			}
+		}
+		h.registered = h.err == nil
+		close(h.done)
+		h.registerMutex.Unlock()
+		return h.err
+	} else {
+		// We already registered. We don't need the lock any more.
+		h.registerMutex.Unlock()
+		if _, err := q.Identify(&h.uuid); err != nil {
+			return err
+		}
+		if err := q.Create(h.cfg); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (h *QueueConnectionHandler) Deactivated(conn *tarantool.Connection,
+	role connection_pool.Role) error {
+	return nil
+}
+
+// Example demonstrates how to use the queue package with the connection_pool
+// package. First of all, you need to create a connection handler for the
+// a ConnectionPool object to process new connections from RW-instances.
+//
+// You need to registry a shared session UUID at a first master connection.
+// It needs to be used to re-identify as the shared session on new
+// RW-instances. See QueueConnectionHandler.Discovered() implementation.
+//
+// After that, you need to create a ConnectorAdapter object with RW mode for
+// the ConnectionPool to send requests into RW-instances. This adapter can
+// be used to create a ready-to-work queue object.
+func Example_connectionPool() {
+	servers := []string{
+		"127.0.0.1:3014",
+		"127.0.0.1:3015",
+		"127.0.0.1:3016",
+	}
+	connOpts := tarantool.Opts{
+		Timeout: 500 * time.Millisecond,
+		User:    "test",
+		Pass:    "test",
+	}
+
+	cfg := queue.Cfg{
+		Temporary:   false,
+		IfNotExists: true,
+		Kind:        queue.FIFO,
+		Opts: queue.Opts{
+			Ttl: 10 * time.Second,
+		},
+	}
+	h := NewQueueConnectionHandler("test_queue", cfg)
+	poolOpts := connection_pool.OptsPool{
+		CheckTimeout:      1 * time.Second,
+		ConnectionHandler: h,
+	}
+	connPool, err := connection_pool.ConnectWithOpts(servers, connOpts, poolOpts)
+	if err != nil {
+		log.Fatalf("unable to connect to the pool: %s", err)
+	}
+
+	defer connPool.Close()
+
+	<-h.done
+
+	if h.err != nil {
+		log.Fatalf("unable to identify in the pool: %s", h.err)
+	}
+
+	rw := connection_pool.NewConnectorAdapter(connPool, connection_pool.RW)
+	q := queue.New(rw, "test_queue")
+
+	fmt.Println("A Queue object is ready to work.")
+
+	if _, err = q.Put("test_data"); err != nil {
+		log.Fatalf("unable to put data into the queue: %s", err)
+	}
+
+	if task, err := q.Take(); err != nil {
+		log.Fatalf("unable to take data from the queue: %s", err)
+	} else {
+		task.Ack()
+		fmt.Println("data:", task.Data())
+	}
+	// Output:
+	// A Queue object is ready to work.
+	// data: test_data
```

Comment:
It would be nice to have some example that checks up a case with master switch.

Reply:
It is not a trivial task. We don't have example how to switch a master in the replicaset (and replicaset configuration too =) connection_pool in tests are just unlinked instances), this may complicate the example too much (and its essence will be lost).

But I can try.

Reply:
> and its essence will be lost

Isn't the essence of connection pool is to provide a way to work with cluster (replicaset) disregarding of what state it has?

Reply:
The connection pool does not have failover or replicaset logic itself. So the code will be in the example. It will make it much more difficult.

Reply:
Does it at least work manually? Or this example doesn't support failover?

Reply:
It works manually =)

Reply:
Well, ok, let it be.

Reply:
I've tried to run the example and check if switchover works and, well, it doesn't seems like it is:
```
=== RUN   Example_connectionPool
2022/09/20 13:37:22 tarantool: storing connection to 127.0.0.1:3015 canceled: .../.rocks/share/tarantool/queue/abstract/queue_session.lua:223: The UUID 6828cabe-9471-4b56-b99a-35f326c223ee is unknown. (0x20)
2022/09/20 13:37:23 tarantool: storing connection to 127.0.0.1:3015 canceled: .../.rocks/share/tarantool/queue/abstract/queue_session.lua:223: The UUID 6828cabe-9471-4b56-b99a-35f326c223ee is unknown. (0x20)
2022/09/20 13:37:52 unable to take data from the queue: can't find rw instance in pool
FAIL	github.com/tarantool/go-tarantool/queue	122.469s
FAIL
```

It's most likely that I had misconfigured something (I messed with `queue/config.lua` to setup replication -- and, to be honest, it wasn't an easy task), but it is a bad signal: if user really want to work with queue, he need to take the example and complete it so it would really be useful. Because example in its current form is not so useful: it demonstrates how to work with single node queue with connection pool (which is a bit weird).

So, once again, I want to raise a question about having a proper test (at least in the form of an example).

Reply:
The example is how to use queue subpackage with a connection_pool subpackage.

It is not about ConnectionPool configuration or replicaset configuration or failover configuration. It is just:

1. Set up Connection handlers for ConnectionPool.
2. Use ConnectionPool.ConnectorAdapter().
3. Use the adapter for a queue.

Reply:
I've added a sample files for manual play with the code. Please, see an update.

Reply:
> The example is how to use queue subpackage with a connection_pool subpackage.

Such example is useless if the only thing it does is work with single-node queue.

Reply:
> I've added a sample files for manual play with the code. Please, see an update.

It should be easy to switch instance ro state -- just a simple eval. So the only thing we need to implement for real example is to configure pool as replicaset. Since you're already have a draft of `config.lua`, a part of work is done. I don't think it would be too time-consuming.

Reply:
And if it won't be enough to provide an example with switchover, it means that this example is not a good example of working with queue replicaset (since connection_pool is designed to work with replicaset) and user won't be able to use it.

Reply:
> The example shows just how a user may to connect a `queue` with a `connection_pool`.

It would be nice to have an example which user can build its logic on. And by "build its logic on" I don't mean implementing handles for basic behavior like failover.

If a new user would want to work with fail-tolerant queue based on Tarantool from Go, I assume that he won't be able to do it unless he spend a lot of time studying both Tarantool and go-tarantool. The main reason I think so is because I couldn't do it myself in short span of time (and I have both Tarantool and go-tarantool experience before) and I don't think that starting a full-fledged debug cycle is a good feature of an example.

Reply:
And, well, it still isn't nice to don't have a test of a feature if the feature is complicated. And this one is complicated. I'm okay of using example as a test since it is a test in some way (`go test` runs it and checks its output), but I not support ignoring integration test since tarantool/queue never will have one with go-tarantool too.

Reply:
> since connection_pool is designed to work with replicaset

It actually does not. It is designed to work with a set of instances. It does have any logic to work with replicaset.

It can not start an instance, it can not configure an instance, it can not change a role of an intance. The repository is a connector, we provide basic blocks for a feature implementation.

Reply:
When we've discussed with Product team the idea of connection pools, in result we've came to a conclusion that it must be able to work with replicaset. Working with vshard cluster is bothersome and rather useless since you can (and should) just implement procedures on vshard router. By using connection pool, you may use the same API as in a single node connection, but with fault-tolerant replicaset underneath. While working with data, it makes any sense only in case when write and read operations results are the same. Otherwise, you'll need to play Russian Roulette (considering underlying round-robin nature) with `get(key)` until you find an instance that contains a record with key. `connection_pool` with plain set of unrelated instances is useless in most real cases.

Reply:
>  but I not support ignoring integration test since tarantool/queue never will have one with go-tarantool too.

What is the idea of your integration test? What do you propose to test? Complex tasks can be splitted into subtasks.

1. We don't need the `ConnectionPool` to check is the `Queue` could work on a new master instance.
2. We don't need the `Queue` to check is the `ConnectionPool` handles connection properly (we already have such tests).
3. We even don't need the `ConnectionPool` to test the handlers code of the example.

Reply:
> If a new user would want to work with fault-tolerant queue based on Tarantool from Go, I assume that he won't be able to do it unless he spend a lot of time studying both Tarantool and go-tarantool. The main reason I think so is because I couldn't do it myself in short span of time (and I have both Tarantool and go-tarantool experience before) and I don't think that starting a full-fledged debug cycle is a good feature of an example.

It depends on what we want to see from the connector: universal building blocks or the implementation of some behavior. 
The current example just shows one of the ways to connect the building blocks. No more than that. I didn't envisage it as only possible one way of the connection.

Reply:
> What is an idea of your integration test? What do you propose to test? Complex tasks can be splitted into subtasks.

1. We connect to a queue replicaset.
2. User takes a task.
3. Master had switched
4. User acknowledges a task.

Reply:
> Complex tasks can be splitted into subtasks.

It sounds like a case when one forgets to write an integration test because they already have unit tests.

Reply:
I believe that everything is work separately. But it is tricky to take these bricks and make them work together. When I took existing example and tried to work with queue replicaset (and I had seen that example sets up queue cfg to work with replicaset), it didn't worked out.

This PR claims that queue work with connection pool. If connection pool is an unrelated set of instances, one may work with queue in it only as a single node connection_pool, which is rather absurd. It is expected (at least from me, and I don't think I'm the only one) that queue replicaset works with connection pool.

Reply:
> It depends on what we want to see from the connector: universal building blocks or the implementation of some behavior. 

User could built queue with connection pool even before this PR, if he really wanted to. ConnectionAdapter made this quest a bit easier. Since we've introduced ConnectionAdapter building block, we need some example on how one should work with it. And we're already have an idea of such example: work with queue replicaset. It's mostly because of this example we have introduced https://github.com/tarantool/go-tarantool/pull/210 and https://github.com/tarantool/go-tarantool/pull/206 . And this PR should finish the story. 

> The current example just shows a one way to connect the building blocks. No more than that. I didn't envisage it as only possible way of the connection.

Failover is a basic behavior. If someone wouldn't want it, he just may ignore the example.

We either show user how to work with queue replicaset or we don't. And if we show, it better be reproducible.

Reply:
This discussion takes more time than needed to actually finish the test.